src/librustc_mir/borrow_check/nll/constraints/graph.rs

```diff
@@ -103,27 +104,57 @@ impl<D: ConstraintGraphDirecton> ConstraintGraph<D> {
     }
 
     /// Given a region `R`, iterate over all constraints `R: R1`.
-    crate fn outgoing_edges(&self, region_sup: RegionVid) -> Edges<'_, D> {
+    crate fn outgoing_edges<'a>(
+        &'a self,
+        region_sup: RegionVid,
+        constraints: &'a ConstraintSet,
+    ) -> Edges<'a, D> {
         let first = self.first_constraints[region_sup];
         Edges {
             graph: self,
+            constraints,
             pointer: first,
+            next_static_idx: None,
         }
     }
 }
 
 crate struct Edges<'s, D: ConstraintGraphDirecton> {
     graph: &'s ConstraintGraph<D>,
+    constraints: &'s ConstraintSet,
     pointer: Option<ConstraintIndex>,
+    next_static_idx: Option<usize>,
 }
 
 impl<'s, D: ConstraintGraphDirecton> Iterator for Edges<'s, D> {
-    type Item = ConstraintIndex;
+    type Item = OutlivesConstraint;
 
     fn next(&mut self) -> Option<Self::Item> {
         if let Some(p) = self.pointer {
-            self.pointer = self.graph.next_constraints[p];
-            Some(p)
+            if let Some(next_static_idx) = self.next_static_idx {
+                if next_static_idx == (self.graph.first_constraints.len() -1) {
+                    self.next_static_idx = None;
+                    self.pointer = self.graph.next_constraints[p];
+                } else {
+                    self.next_static_idx = Some(next_static_idx + 1);
+                }
+
+                Some(OutlivesConstraint {
+                    sup: 0.into(),
+                    sub: next_static_idx.into(),
+                    locations: Locations::All,
+                })
+            } else {
+                self.pointer = self.graph.next_constraints[p];
+
+                let constraint = self.constraints[p];
+
+                if constraint.sup == 0.into() {
```

Comment:
Hmm, I'm not crazy about hardcoding `0` here =) 

Two options:

- Either take the `UniversalRegions` as a argument somewhere
- Or refactor `UniversalRegions` code to have a `FR_STATIC` constant

Reply:
But also, I think we may have to consider the direction here...? e.g., if we this is a "normal" graph, then this behavior seems right (because outgoing edges come from the "sup"). But if this is a "reverse" graph, then I think it's not right. Given a graph like

```
'a: 'b
'static: 'b
```

If you ask for the outgoing edge from `'b`, you're going to get `'a`, `'static`... and then an edge to every other region, right?

TBH, I think in a reverse graph we can just ignore the "static" case. I mean, to be strictly consistent, we would basically add an "extra" static constraint to every region in a reverse graph, but it wouldn't actually matter to any of the consumers.

Reply:
One final note is that this current setup will first walk the edges of the graph and *then* walk the static successors. I think this results in some of the "duplicate" errors below. Putting all this together, maybe it'd be easier to modify the `edges` function so that it checks whether:

- the starting node is `FR_STATIC`
- you are in normal mode (just make the `Direction` trait have a boolean `is_normal()` fn or something)

and, if both are true, set `next_static_idx` to `Some(0)` and `pointer` to `None`.

Reply:
What do you think?